### PR TITLE
MSVC Build: Support VS2017 or VS2015 compiler tools

### PR DESCRIPTION
**Pull request by jeffhostetler**
Teach Makefile to use whatever VS compiler tool chain
is installed on the system.

Previously I had hard-coded paths to VS2015.  This change
uses the new VsWhere.exe and VsDevCmd.bat in a background
Command Prompt process to compute the proper environment
variables for bash and make.

If VS2017u2 or higher is installed, it will use the
compiler, linker, libaries associated with it.

It will fall back to the previous configuration using
VS2015.  It should be fairly easy to extend the fallback
section in the "find_vs_env.bat" script to support older
compilers.

I have tested this with VS2017.3 (Dev15.3) and the V140
command line tools (distributed with VS2015).

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

Thanks for taking the time to contribute to Git! Please be advised that the
Git community does not use github.com for their contributions. Instead, we use
a mailing list (git@vger.kernel.org) for code submissions, code reviews, and
bug reports. Nevertheless, you can use submitGit to conveniently send your Pull
Requests commits to our mailing list.

Please read the "guidelines for contributing" linked above!

### CODE REVIEW

File: config.mak.uname [hunk 1]
```diff
@@ -17,6 +17,13 @@ ifdef MSVC
 	uname_O := Windows
 endif
 
+
+ifdef MSVC
+GIT-MSVC-GEN: ./compat/vcbuild/find_vs_env.bat
+	@./compat/vcbuild/find_vs_env.bat | sed 's|\\|/|g' >GIT-MSVC-GEN
+-include GIT-MSVC-GEN
+endif
+
```
Comment (dscho) [inline config.mak.uname]: Let's move this inside the already-existing clause?

Comment (dscho) [inline config.mak.uname]: Also: whose duty is it to generate `GIT-MSVC-GEN`? I *think* you want something like `Makefile` does for `GIT-VERSION` instead (`FORCE` as dependency, then test in the `.bat` whether it actually needs to run).

Comment (jeffhostetler) [inline config.mak.uname]: @dscho From what I can tell, a generated included file causes an implicit dependency.  Make automatically takes care of forcing the file to be generated.

File: compat/vcbuild/find_vs_env.bat [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+REM You can use either GCC (the default) or MSVC to build git
+REM using the GIT-SDK command line tools.
+REM        $ make
+REM        $ make MSVC=1
+REM
+REM GIT-SDK BASH windows inherit environment varibles with all of
+REM the bin/lib/include paths for GCC.  It DOES NOT inherit values
+REM for the corresponding MSVC tools.
+REM
+REM During normal (non-git) Windows development, you launche one
```
Comment (danilaml) [inline compat/vcbuild/find_vs_env.bat]: > you launche one

typo

File: compat/vcbuild/find_vs_env.bat [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+REM
+REM ================================================================
+REM Note: Throughout this script we use "dir <path> && <cmd>" rather
+REM than "if exist <path>" because of script problems with pathnames
+REM containing spaces.
+REM ================================================================
+
+REM Sanitize PATH to prevent git-sdk paths from confusing "wmic.exe"
+REM (called internally in some of the system BAT files).
+SET PATH=C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;
```
Comment (danilaml) [inline compat/vcbuild/find_vs_env.bat]: Sorry if I misunderstood something but is this scripted is to be used by the git developers (not build bots)?
If so, It doesn't look like a good idea to hard-code drive C everywhere. For instance, on my PC windows is installed on drive `G` and IRC starting from VS2017 visual studio is not bound to a system drive either (I'm not sure whether vswhere is still installed to the default  `%programfiles(x86)%`) .

**Comment by mfriedrich74** [comment]
Each Visual Studio version creates an environment variable pointing
somewhere  into its installation folder. That can be used derive correct
settings or to call its vcvars.bat.

Best regards, Mike

**Comment by danilaml** [comment]
@mfriedrich74 doesn't seem to be the case starting from VS2017 and going forward (if I understood [this](https://blogs.msdn.microsoft.com/vcblog/2017/03/06/finding-the-visual-c-compiler-tools-in-visual-studio-2017/) correctly).

File: config.mak.uname (hunk 1 above)
Comment (dscho) [inline config.mak.uname]: I see! Thanks for the explanation.

**Comment by dscho** [comment]
Thanks all, in particular @jeffhostetler for the original work and @danilaml for the thorough and very helpful review!

**Comment by jeffhostetler** [comment]
@mfriedrich74  Yes, it is a bit confusing.  The layout of VS2017 (and beyond) is different from the layout of VS2015 which was different from earlier layouts.  Yes, the stock install does/can set an environment variable pointing at the installation directory, or adds the directory containing the vcvars to the PATH, but we want to go a little deeper here and get the proper values for the "-I" and "-L" args to CL.exe.  Those change between versions.  For example, VS2015 introduced the UCRT and Windows Kits changes, so builds based on it or newer, have different values from previous generations.

So my script tries to find the most recent version of VS installed and build the version-specific compiler flags.

I left the TODO at the bottom because I cannot test the earlier versions (I don't have any of them installed).  Also, I've not built it with the earlier tools and there might be compiler errors and etc.,
so I didn't just want to assume it would work.  The existing TODO and overall layout of the script
should make it easy for someone to extend it if/when necessary.

Thanks!

**Comment by dscho** [comment]
> ... because I cannot test the earlier versions (I don't have any of them installed).

@PhilipOakley maybe you are interested?

**Comment by PhilipOakley** [comment]
> > ... because I cannot test the earlier versions (I don't have any of them installed).

> @PhilipOakley maybe you are interested?

While I'm interested, unfortunately I'm rather time poor at the moment, 

I will add it as part of my long todo list, but don't hold your breath just yet.

**Pull request merged**